Updates VideoDecoder plugin API to GDExt.

zombietype opened the pull request:
Adds VideoStreamExtension and relevant resource loaders to handle external GDNative plugins.

There are known bugs in godot-cpp w.r.t. reference counting in Ref<T> (https://github.com/godotengine/godot-cpp/issues/652)
However, from godot's side this should work just fine once that is fixed.

TODO:
Integrate Theora to use VideoDecoderServer.

*Bugsquad edit: This closes https://github.com/godotengine/godot-proposals/issues/3286.*

kidrigger commented on modules/gdnative/SCsub:
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env python
```
Mistake. Will clear in another commit.

fire commented:
There are some cicd errors.

akien-mga commented:
Needs a rebase to fix CI.

akien-mga commented on main/main.cpp:
```diff
@@ -1810,6 +1812,9 @@ Error Main::setup2(Thread::ID p_main_tid_override) {
 	audio_server = memnew(AudioServer);
 	audio_server->init();
 
+	// init the Video decoder server
+	video_decoder_server = memnew(VideoDecoderServer);
```
Does it actually need to be this early? For XR it's a special case as OpenXR needs to be set up before Vulkan, but I don't think that matters for video decoding?

kidrigger replied:
It used to be important when I had to set it up before the plugins loaded. It doesn't matter anymore. Where should I put it?

Actually I can just turn the whole class static since pretty much everything that matters is.

reduz commented:
We are not using the Extension class much, I suggest checking how this is done for AudioStream, including how the class is registered as VIRTUAL. Other than that I think its fine.

zombietype committed:
Updates VideoDecoder plugin API to GDExtension.

Adds VideoStream and relevant resource loaders to migrate
external GDNative plugins to GDExtension.

Adds a VideoStreamLoader as a specialization of ResourceFormatLoader
as ClassDB::is_parent_class is inaccessible from GDExtension currently.

Using Object* instead of Ref<T> in order to avoid the refcount bug
(godotengine/godot-cpp#652)
Also another bug is in ResourceLoader in use on the extension side that
requires fixing.

reduz commented on servers/video_decoder_server.h:
```diff
@@ -1,5 +1,5 @@
 /*************************************************************************/
-/*  video_stream.h                                                       */
+/*  video_decoder_server.h                                               */
```
I think a VideoDecoderServer is for the most part unnecesary, with GDExtension, it should simply possible to provide the VideoStream<decodername> with the proper ResourceFormatLoader, both using GDExtension and that should be enough.

kidrigger replied:
It's meant to bypass the lack of ClassDB features. There's no way to check ClassDB::is_parent_class.
And I don't know how to add that either.

reduz replied:
ClassDB should be exposed on the GDExtension side, what check do you need to do?

kidrigger replied:
```ClassDB::is_parent_class(p_type, "VideoStream");```
The `is_parent_class` function is not exposed.
Secondly everything hinges on https://github.com/godotengine/godot-cpp/issues/652 being fixed. Currently the plugins all crash on exit due to double freeing.

kidrigger replied:
Also had a hard time registering new video formats from GDExtension. Possibly as a consequence of the Ref bug.
Trying to work in the Object* workaround.

kidrigger commented:
Bypasses Ref<T> issue (https://github.com/godotengine/godot-cpp/issues/652) by using Object* when the ownership is completely transferred from the Extension to Godot.

Plugins make the engine crash on exit due to issues with _instance_bindings. Creating issue.

kidrigger commented:
> If you've got a moment, can you write a guide how to test, and how to develop for this? Let me know if it's too much work.

This commit registers all the virtual functions to their closest supported behaviors in GDExtension.

For the WebM plugin, clone the following:
https://github.com/kidrigger/godot-video-reference/tree/webm-integ-serverless

SCons should work with similar options as Godot (target=editor p=x11 ... )
Use target as editor.

Use option test=yes to generate the Dynamic library for the test project (a godot project in the repo itself)
Run godot and load the editor. You should have a scene with a VideoPlayer with a pre-loaded BigBuckBunny Webm video.
If it plays with sound - it works.

You should be able to copy and paste the plugin directory to any project without issues.

Custom plugins for other codecs can register their own classes and decoders in the same way as the reference.

lyuma commented:
Fire is asking if you can update this to fix the conflicting files. It would be great to get this into 4.0

My understanding is there is agreement that it needs to get in place in time for 4.0. Due to the holidays and time constraints at the GDExtension meetings, it didn't get discussed until recently.

If you don't have time to update this, then I can try my hand.

kidrigger commented:
> Fire is asking if you can update this to fix the conflicting files. It would be great to get this into 4.0
> 
> My understanding is there is agreement that it needs to get in place in time for 4.0. Due to the holidays and time constraints at the GDExtension meetings, it didn't get discussed until recently.
> 
> If you don't have time to update this, then I can try my hand.

Updated. Hope it can be merged soon.

lyuma commented:
Hi,
I appreciate that you went ahead and updated the PR... I have no idea what happened but it looks like git tried to include a merge into your rebase.

I went ahead and redid the rebase myself since I'm familiar with git and didn't want you to waste time trying to figure out how to undo that merge. It's all okay.

A quick question: is it okay if I move the video_stream.h/cpp back into resources? I see the analogy with audio, but a lot of times, the Godot approach to organization is pragmatic and it seems a bit overkill to have a folder with three files in it. If we get more video-related code, it could warrant its own directory, but right now I'd prefer to keep things simpler.

kidrigger commented:
> Hi,
> I appreciate that you went ahead and updated the PR... I have no idea what happened but it looks like git tried to include a merge into your rebase.
> 
> I went ahead and redid the rebase myself since I'm familiar with git and didn't want you to waste time trying to figure out how to undo that merge. It's all okay.
Thank you!
> 
> A quick question: is it okay if I move the video_stream.h/cpp back into resources? I see the analogy with audio, but a lot of times, the Godot approach to organization is pragmatic and it seems a bit overkill to have a folder with three files in it. If we get more video-related code, it could warrant its own directory, but right now I'd prefer to keep things simpler.

Sounds good.
I hope this gets merged soon!

lyuma commented:
Hey, so I noticed a bunch of new file apis being added, which aren't used by theora.
For example, VideoStreamPlaybackTheora uses set_file while VideoStreamPlayback uses open_file.

I also noticed VideoStreamPlayback.open_file isn't exposed but is called automatically. additionally new file_seek, file_len, file_pos functions were added which don't seem to be called. Furthermore, they seem to just be wrapping FileAccess. I would like to remove these, and instead call set_file on the gdext... and then if the extension need to open a file, it should it do it by itself.

kidrigger commented:
> Hey, so I noticed a bunch of new file apis being added, which aren't used by theora.
> For example, VideoStreamPlaybackTheora uses set_file while VideoStreamPlayback uses open_file.
> 
> I also noticed VideoStreamPlayback.open_file isn't exposed but is called automatically. additionally new file_seek, file_len, file_pos functions were added which don't seem to be called. Furthermore, they seem to just be wrapping FileAccess. I would like to remove these, and instead call `_set_file` on the gdext instead of `_file_opened`... and then if the extension need to open a file, it should it do it by itself.

Last I remember the whole FileAccess wrapper existed because FileAccess wasn't accessible from GDExtension.

Please make sure FileAccess is exposed to extension before modifying that.

lyuma commented:
Yes, FileAccess was fully exposed to GDExtension sometime around beta3. So that is no longer a concern.

In that case, I'll remove the file_ functions from this PR. Everything else is looking good.
Next step is to test this with your webm implementation and make sure it all works end-to-end.

kidrigger commented:
I won't be able to add the time features in the plugin until after the 8th.

On Mon, Jan 30, 2023, 23:40 lyuma ***@***.***> wrote:

> Yes, FileAccess was fully exposed to GDExtension sometime around beta3. So
> that is no longer a concern.
>
> In that case, I'll remove the file_ functions from this PR. Everything
> else is looking good.
> Next step is to test this with your webm implementation and make sure it
> all works end-to-end.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/62737#issuecomment-1409471902>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEDSN5UN4O6F4LYLE5ICWQDWVA7O3ANCNFSM52WL4X7A>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

lyuma commented on scene/resources/video_stream.h:
```diff
@@ -39,40 +40,103 @@ class VideoStreamPlayback : public Resource {
+	_bind_methods();
+
+	// Using Object* to bypass godot-cpp$652
+	GDVIRTUAL0R(Object *, _instantiate_playback);
```
This should return Ref<VideoStreamPlayback>. I'll fix it.

lyuma replied:
(Since https://github.com/godotengine/godot-cpp/issues/652 was fixed in december)

akien-mga commented:
Thanks!

fire commented:
set_loop and get_loop was removed. @lyuma Is this intended?

lyuma commented:
I removed them because they were not being called by Godot, except in one place which hardcoded `set_loop(false)`.

lyuma commented:
if we add a way to perform set_loop(true) in Godot, we can add the functionality back in. but I wasn't sure the intended usage, so I couldn't properly document it.

GeorgeS2019 commented:
will there be a godot4 demo using this API?

Calinou commented:
> will there be a godot4 demo using this API?

https://github.com/kidrigger/godot-video-reference is pretty much that demo :slightly_smiling_face:
